orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/tasks/quip/TriggerQuipTask.groovy

```diff
@@ -31,42 +33,88 @@ import retrofit.client.Client
-        String hostName = valueMap.hostName
-        def instanceService = createInstanceService("http://${hostName}:5050")
-
-        try {
-          def instanceResponse = instanceService.patchInstance(packageName, version, "")
-          def ref = objectMapper.readValue(instanceResponse.body.in().text, Map).ref
-          taskIdMap.put(hostName, ref.substring(1 + ref.lastIndexOf('/')))
-        } catch (RetrofitError e) {
-          log.warn("Error in Quip request: {}", e.message)
-          executionStatus = ExecutionStatus.RUNNING
+      remainingInstances.each { String instanceId, Map instance ->
+        String instanceHostName = instance.hostName
+        def instanceService = createInstanceService("http://${instanceHostName}:5050")
+        if (stage.context.skipUpToDate &&
+          // optionally check the installed package version and skip if == target version
+          (getAppVersion(instanceService, packageName) == version)) {
+            skippedInstances.put(instanceId, instance)
+            //remove from instances so we don't wait for it to quip
+            instances.remove(instanceId)
+        } else {
+          try {
+            def instanceResponse = instanceService.patchInstance(packageName, version, "")
+            def ref = objectMapper.readValue(instanceResponse.body.in().text, Map).ref
+            taskIdMap.put(instanceHostName, ref.substring(1 + ref.lastIndexOf('/')))
+            patchedInstanceIds << instanceId
+          } catch (RetrofitError e) {
+            log.warn("Error in Quip request: {}", e.message)
+          }
         }
       }
+      remainingInstances.keySet().removeAll(patchedInstanceIds)
+      remainingInstances.keySet().removeAll(skippedInstances.keySet())
     } else {
       throw new RuntimeException("one or more required parameters are missing : version (${version}) || package (${packageName})|| instances (${instances})")
     }
-    stageOutputs.put("taskIds", taskIdMap)
-    return new DefaultTaskResult(executionStatus, stageOutputs)
+    Map stageOutputs = [
+      taskIds: taskIdMap,
+      instances: instances,
+      instanceIds: instances.keySet(), // for WaitForDown/UpInstancesTask
+      skippedInstances: skippedInstances,
+      remainingInstances: remainingInstances,
+      version: version
+    ]
+    return new DefaultTaskResult(remainingInstances ? ExecutionStatus.RUNNING : ExecutionStatus.SUCCEEDED, stageOutputs)
+  }
+
+  String getAppVersion(InstanceService instanceService, String packageName) {
+    int retries = 5;
+    def instanceResponse
+    String version
+
+    while (retries) {
+      try {
+        instanceResponse = instanceService.getCurrentVersion(packageName)
+        version = objectMapper.readValue(instanceResponse.body.in().text, Map)?.version
+        if (version && !version.isEmpty()) {
+          return version
+        }
+      } catch (RetrofitError e) {
+        //retry
+      }
+      sleep(INSTANCE_VERSION_SLEEP)
```

Comment:
not super happy about this but it mirrors what was happening at stage evaluation time
ideally this would be its own retryable task